Updated README and example projects

duartepinto opened the pull request:
In this PR I add two example projects (one in Swift and another in Objc) and also update the README with examples from said example projects.

Both applications only have a button, to force a tracking request, and a text label to show the result of said tracking request.

This PR also includes:
* Fix codecov reports by ignoring Pod files
* Use xcpretty in travis

duartepinto committed:
Travis test script uses main scheme

duartepinto committed:
Init 'Examples' project

duartepinto committed:
Make track request

duartepinto committed:
Start VelocidiSDK after application launch

duartepinto committed:
Add missing local Pods file

duartepinto committed:
Make tracking request and print result to label

duartepinto committed:
Remove local Pods from Examples project

duartepinto committed:
Use xcpretty in travis.
Use VelocidiSDKTests no longer a shared target.
Remove use of deprecated function.
Fix code coverage reports to codecov configurations.

Remove additional SDK  'i' warning

duartepinto committed:
Add Installation and Usage to README

duartepinto committed:
Rename Examples project to SwiftExample

duartepinto committed:
Change Podfile and Pod paths

duartepinto committed:
Add Objective-C example project

duartepinto committed:
Add Pods to ObjcExample. Tracking event in ObjcExample

duartepinto committed:
Add Objective-C examples to README. Small fix in ObjcExample

duartepinto committed:
Add README for Examples

duartepinto committed:
Use newer cocoapods version

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,157 @@
+
+### Send a track event
+
+In order to sent a tracking event, create an instance of TrackingEvent. Then, call the singleton instance of `VSDKVelocidi` and use the method `track`
```
```suggestion
In order to send a tracking event, create an instance of `VSDKTrackingEvent`. Then, call the singleton instance of `VSDKVelocidi` and use the `track` method.
```

antonio-ramadas commented on README.md:
````diff
@@ -8,3 +8,157 @@
+```
+
+
+You can also pass callback blocks that will that will be called when the request is successful or if if fails.
````
```suggestion
You can also pass callback blocks that will be called when the request either succeeds or fails.
```

antonio-ramadas commented on README.md:
````diff
@@ -8,3 +8,157 @@
+}];
+```
+
+#### Available tracking events model classes
````
```suggestion
#### [Available tracking events model classes](https://github.com/velocidi/velocidi-ios-objc-sdk/tree/master/VelocidiSDK/VelocidiSDK)
```

duartepinto replied:
Don't really like this. I think it's strange to have a link in a README header. If I did not know I would assume it would be a link to copy the link, similar to this 
![image](https://user-images.githubusercontent.com/7612420/58255096-1c821700-7d64-11e9-9bf4-e392336d3905.png)

antonio-ramadas replied:
Hum... I would not have that impression. Anyway, with the list below I don't find it necessary to have a link there. It would be a "nice to have".

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,157 @@
+<!-- #### Create your custom tracking event -->
+<!-- If none of the available tracking events matches your needs you can also extend `TrackingEvent` and create your own.  -->
+
+<!-- _Beware! Custom tracking events are not interpreted by our services and bring limitied functionality other than logging and basic statistics._ -->
```
```suggestion
<!-- _Beware! Custom tracking events are not interpreted by our services and bring limited functionality other than logging and basic statistics._ -->
```

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,157 @@
+* VSDKSearch
+* VSDKSubscription
+
+<!-- #### Create your custom tracking event -->
```
Is this supposed to stay commented?

duartepinto replied:
Either commented or deleted. Ran into troubles when doing this. We cannot use JSONModel, the framework we use to serialize classes to JSON, in Swift and therefore we cannot make custom tracking events.

antonio-ramadas replied:
What alternatives a developer has to send a custom event?

duartepinto replied:
If he uses Objective-C, inherit `VSDKTrackingEvent` or any other class that also inherits from `VSDKTrackingEvent`. If he uses Swift, none.

antonio-ramadas replied:
Can't we implement a new method that receives a dictionary and sends it as JSON?

duartepinto replied:
Maybe, I'd have to look at it. But ideally, we would not have to move away from inheritance or from using objects. Either way, I just left this here because I ended up forgetting I had commented it (this does not appear when the .md is rendered so I forgot that I had written it). Meant to leave this for a proper discussion about what we should do:
* Document how to use custom events in Objective-C and don't offer this functionality in Swift.
* R&D how to properly fix this, maybe by changing to a new JSON library.
* Avoid the problem by doing something like what you suggested.

antonio-ramadas replied:
If I have a Swift project, can I write an Objective-C class and use it in Swift?

duartepinto replied:
Should work, yes.

antonio-ramadas replied:
At least for now, that seems good enough for me. What do you think? Can you document it?

duartepinto replied:
Yes, I'll look into it. It's not a very good approach and we should revise this in the future, but I think that for now is the best we have.

antonio-ramadas replied:
I noticed you re-requested my review. Has this been fixed?

duartepinto replied:
Sorry, my bad 😅

duartepinto replied:
Please take a look at 9bf9e78 d201da1 18be956 and d394ec1

antonio-ramadas commented on Examples/README.md:
````diff
@@ -0,0 +1,17 @@
+$ while true; do echo -e 'HTTP/1.1 200 OK\r\n\r\n' | nc -i 1 -l 8080; done
+```
+
+_Netcat has limited functionality and this command will wait 1 second before replying(in order to wait for all data to be transmitted ), so the it might feel a little slugish. You can also use online mock HTTP clients, like [Mocky](https://www.mocky.io/)_ 
````
```suggestion
_Netcat has limited functionality and this command will wait 1 second before replying (in order to wait for all data to be transmitted). To overcome this delay you can use any other tool, including online mock HTTP clients like [Mocky](https://www.mocky.io/)._ 
```

duartepinto replied:
I feel like this is saying the same thing just in a different way.

antonio-ramadas replied:
Yes, it is. It is a nitpick of mine (I don't like "it might feel a little slugish"). I don't have strong feelings. However, if you ignore, please fix spacing and spelling:

```suggestion
_Netcat has limited functionality and this command will wait 1 second before replying (in order to wait for all data to be transmitted), so it might feel a little sluggish. You can also use online mock HTTP clients, like [Mocky](https://www.mocky.io/)._ 
```

antonio-ramadas commented on Examples/README.md:
````diff
@@ -0,0 +1,17 @@
+
+To see the HTTP events sent by the SDK, have the configuration url's point to `http://localhost:8080` and then run a netcat instance that replies to all request with a `200 OK` response code.
+
+```
````
```suggestion
```bash
```

antonio-ramadas commented on Examples/ObjcExample/ObjcExample/ViewController.m:
```diff
@@ -0,0 +1,31 @@
+    trackingNumber++;
+    
+    [VSDKVelocidi.sharedInstance track: trackingEvent onSuccess: ^(NSURLResponse * response, id responseObject){
+        self.mainLabel.text = [NSString stringWithFormat: @"Tracking request #%i successful!!!", trackingNumber];
```
```suggestion
        self.mainLabel.text = [NSString stringWithFormat: @"Tracking request #%i successful!", currentTrNumber];
```

antonio-ramadas commented on Examples/ObjcExample/ObjcExample/ViewController.m:
```diff
@@ -0,0 +1,31 @@
+    [VSDKVelocidi.sharedInstance track: trackingEvent onSuccess: ^(NSURLResponse * response, id responseObject){
+        self.mainLabel.text = [NSString stringWithFormat: @"Tracking request #%i successful!!!", trackingNumber];
+    } onFailure: ^(NSError * error){
+        self.mainLabel.text = [NSString stringWithFormat: @"Error with tracking request #%i.\n Error: %@", trackingNumber, [error localizedDescription]];
```
```suggestion
        self.mainLabel.text = [NSString stringWithFormat: @"Error with tracking request #%i.\n Error: %@", currentTrNumber, [error localizedDescription]];
```

antonio-ramadas commented on Examples/ObjcExample/ObjcExample/ViewController.m:
```diff
@@ -0,0 +1,31 @@
+    trackingEvent.clientId = @"foo";
+    trackingEvent.siteId = @"bar";
+    
+    trackingNumber++;
```
```suggestion
    int currentTrNumber = trackingNumber++;
```

antonio-ramadas commented on Examples/SwiftExample/SwiftExample/ViewController.swift:
```diff
@@ -0,0 +1,30 @@
+    //MARK: Actions
+    @IBAction func TrackingEvent(_ sender: UIButton) {
+        let trackingEvent = VSDKPageView()
+        trackingEvent.siteId = "0"
```
```suggestion
        trackingEvent.siteId = "foo"
```

antonio-ramadas commented on Examples/SwiftExample/SwiftExample/ViewController.swift:
```diff
@@ -0,0 +1,30 @@
+    @IBAction func TrackingEvent(_ sender: UIButton) {
+        let trackingEvent = VSDKPageView()
+        trackingEvent.siteId = "0"
+        trackingEvent.clientId = "0"
```
```suggestion
        trackingEvent.clientId = "bar"
```

antonio-ramadas commented on Examples/SwiftExample/SwiftExample/ViewController.swift:
```diff
@@ -0,0 +1,30 @@
+        let currentTrNumber = trackingNumber
+        
+        VSDKVelocidi.sharedInstance().track(trackingEvent, onSuccess:{ (response: URLResponse, responseObject: Any) in
+            self.mainLabel.text = "Tracking request #\(currentTrNumber) successful!!!"
```
```suggestion
            self.mainLabel.text = "Tracking request #\(currentTrNumber) successful!"
```

antonio-ramadas commented on Examples/README.md:
```diff
@@ -0,0 +1,17 @@
+
+These two examples projects illustrate how VelocidiSDK can be used in a project. ObjcExample is written in Objective-C and SwiftExample is written in Swift.
+
+## Instructions
```
As discussed offline, can you improve this section?

duartepinto replied:
Check b3fbf26

antonio-ramadas commented on Examples/Podfile.lock:
```diff
@@ -0,0 +1,41 @@
+
+PODFILE CHECKSUM: ba8278e89e7a7f65f231b280ae9ba6dfdc615d4a
+
+COCOAPODS: 1.6.2
```
Cocoapods is currently at 1.7.0. Can we generate the lock file with the latest version?

duartepinto replied:
Updated it in 5aa7b31

duartepinto committed:
Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update Examples/SwiftExample/SwiftExample/ViewController.swift

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update Examples/SwiftExample/SwiftExample/ViewController.swift

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update Examples/SwiftExample/SwiftExample/ViewController.swift

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Clarify Example's README

duartepinto committed:
Use Cocoapods 1.7.0 in Examples

duartepinto committed:
Update Examples/ObjcExample/ObjcExample/ViewController.m

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update Examples/README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update Examples/ObjcExample/ObjcExample/ViewController.m

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update Examples/ObjcExample/ObjcExample/ViewController.m

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

antonio-ramadas commented on Examples/ObjcExample/ObjcExample/ViewController.m:
```diff
@@ -0,0 +1,31 @@
+    trackingEvent.clientId = @"foo";
+    trackingEvent.siteId = @"bar";
+    
+    int currentTrNumber = trackingNumber++;
```
My bad.

```suggestion
    int currentTrNumber = ++trackingNumber;
```

duartepinto committed:
Update Examples/ObjcExample/ObjcExample/ViewController.m

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Update Examples/README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

duartepinto committed:
Add custom tracking event to ObjC Example

duartepinto committed:
Update README with Custom events

duartepinto committed:
Add Custom Tracking event in Swift Examples

duartepinto committed:
Update README with Swift custom events

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,239 @@
+
+@interface CustomEvent : VSDKTrackingEvent
+
+// Place bellow the desired fields you want to add to the tracking event
```
```suggestion
// Place below the desired fields you want to add to the tracking event
```

[bellow](https://en.oxforddictionaries.com/definition/bellow)
> (of a person or animal) emit a deep loud roar, typically in pain or anger.

-----

[below](https://en.oxforddictionaries.com/definition/below)
> At a lower level or layer than.

antonio-ramadas commented on README.md:
````diff
@@ -8,3 +8,239 @@
+@end
+```
+
+__CustomEvent.h__
````
```suggestion
__CustomEvent.m__
```

antonio-ramadas commented on README.md:
````diff
@@ -8,3 +8,239 @@
+@end
+```
+
+This new class can then be imported and used like any other tracking event
````
```suggestion
This new class can then be imported and used like any other tracking event:
```

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,239 @@
+#### Create your custom tracking event
+If none of the available tracking events matches your needs you can also extend `TrackingEvent` and create your own. 
+
+__Beware!__ _Custom tracking events are not interpreted by our services and bring limited functionality other than logging and basic statistics._
```
They **may** be interpreted. What if I am extending one of the current events? I see your point, but this can be interpreted differently by someone reading this for the first time.

duartepinto replied:
Please check 2f87b04

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,239 @@
+
+Create a new Cocoa Touch class that inherits from `VSDKTrackingEvent` (or any of the other tracking event model classes)
+
+Example:
```
What do you think about having `Example` as a subsection of `Objective-C`?

duartepinto replied:
Don't really have an opinion. But take into consideration that it will be an `<h6>` and hard to differentiate from Objective-C

antonio-ramadas commented on README.md:
````diff
@@ -8,3 +8,239 @@
+trackingEvent.customField = "baz"
+```
+
+If you had any problem with importing the Objective-C class into Swift, please take a look at Apple's guide on [Importing Objective-C into Swift](https://developer.apple.com/documentation/swift/imported_c_and_objective-c_apis/importing_objective-c_into_swift)
````
```suggestion
If you had any problem with importing the Objective-C class into Swift, please take a look at Apple's guide on [Importing Objective-C into Swift](https://developer.apple.com/documentation/swift/imported_c_and_objective-c_apis/importing_objective-c_into_swift).
```

antonio-ramadas commented on README.md:
````diff
@@ -8,3 +8,239 @@
+1. Use the newly created class as any other tracking event model class:
+```swift
+let trackingEvent = CustomEvent()
+trackingEvent.siteId = "foo"
````
For consistency, keep the same from above

```suggestion
trackingEvent.siteId = "RandomSiteId"
```

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,239 @@
+let trackingEvent = CustomEvent()
+trackingEvent.siteId = "foo"
+trackingEvent.clientId = "bar"
+trackingEvent.customField = "baz"
```
For consistency, keep the same from above

```suggestion
trackingEvent.customField = "RandomCustomField"
```

antonio-ramadas commented on README.md:
````diff
@@ -8,3 +8,239 @@
+```swift
+let trackingEvent = CustomEvent()
+trackingEvent.siteId = "foo"
+trackingEvent.clientId = "bar"
````
For consistency, keep the same from above

```suggestion
trackingEvent.clientId = "RandomClientId"
```

antonio-ramadas commented on README.md:
```diff
@@ -8,3 +8,239 @@
+
+##### Objective-C
+
+Create a new Cocoa Touch class that inherits from `VSDKTrackingEvent` (or any of the other tracking event model classes)
```
```suggestion
Create a new Cocoa Touch class that inherits from `VSDKTrackingEvent` (or any of the other tracking event model classes).
```

antonio-ramadas reviewed (changes requested):
Great job with the Swift tutorial on how to create a custom event 👏 

Most (if not all) changes required are all pretty small.

duartepinto committed:
Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>

Update README.md

Co-Authored-By: António Ramadas <antonio_ramadas@hotmail.com>
